rfc/mediachain-rfc-4.md

```diff
@@ -0,0 +1,653 @@
+transfer, such as IPFS Bitswap or Bittorrent.
+
+Similar dynamics arise for the use case of firehose sources, which
+generate a constant load of new statements, eg by following social
+media feeds.  Firehoses don't have to publish every statement in real
+time. Instead, they can buffer statements and periodically publish
+them together with their metadata as archives. This allows the system
+to push high volumes of data without overloading the network.
+
+### Metadata Persistence
+
+The flow of metadata in the system follows the path from sources to
+readers through IPFS. When a reader receives a statement and adds it
+to its local store, it also fetches the associated metadata. Client
+readers are not expected to reseed the object in IPFS, especially if
+they operate in a disconnected fashion.
+
+In order to avoid overloading the source with IPFS requests and
+provide availability even when the source is offline, the metadata is
+reseeded by caches who follow the relevant namespaces. Longer term
+metadata persistence in the form of archive files is provided by
+archivers. These basic services will be supported by organizations
+using the mediachain per normal course of operation.
+
+The issue of persistent seeding in IPFS is a little more complicated.
+Firstly, a seeder must consume memory, storage, and egress bandwidth
+in order to serve the objects. Secondly, IPFS publishing creates
+system traffic in the form of DHT provider announcements. Thus, a form
+of incentives needs to be devised to encourage and justify persistent
+IPFS seeding for high value namespace. This can be achieved with an
+application-specific coin like filecoin or through a well established
+blockchain.
+
+### Indexes and Queries
+
+In order to query the database, a peer can construct a local index by
+processing statements in the local store. This index provides the
+local view of the mediachain, and allows retrieval of statements and
+metadata objects based on multiple criteria.
+
+A basic indexing scheme, which performs only statement envelope processing
+and doesn't parse the metadata, allows retrieval with the following
+criteria:
+
+- Query by WKI, publisher or peer id, statement id, object id
+- Query by namespace(s)
+- Query by timestamp
+
+This baseline index, which can be easily implemented using an embedded
+db like sqlite, is the simplest way to implement polling for state
+synchronization. It also provides a low level interface for advanced
+users, who are free to develop and share more advanced indexing schemes
+that parse object content.
+
+Queries do not have to be limited to one's local view however. For
+instance, a user may want to query joining on namespaces it is not
+currently following.  This could be because the user is exploring the
+namespace or because the particular namespace has a larger index that
+aggregates multiple namespaces. This can be accomplished by directly
+querying other peers or by utilizing more specialized indexers.
```

Comment:
I think we probably want to do queries that join across multiple nodes, but the join can be implemented purely in the client side with stacked queries.